src/contracts/mod.rs

```diff
@@ -1,228 +1,112 @@
-        };
-
-        // TODO: Some way to check the initial leaf
-
-        Ok(Self {
-            ethereum,
-            sitter,
-            semaphore,
-            group_id: options.group_id,
-            tree_depth: actual_tree_depth,
-            initial_leaf: options.initial_leaf,
-        })
-    }
-
-    #[must_use]
-    pub const fn group_id(&self) -> U256 {
-        self.group_id
-    }
-
-    #[must_use]
-    pub const fn tree_depth(&self) -> usize {
-        self.tree_depth
-    }
-
-    #[must_use]
-    pub const fn initial_leaf(&self) -> Field {
-        self.initial_leaf
-    }
-
-    pub async fn confirmed_block_number(&self) -> Result<u64, EventError> {
-        self.ethereum
-            .confirmed_block_number()
-            .await
-            .map(|num| num.as_u64())
-    }
-
-    #[allow(clippy::disallowed_methods)] // False positive from macro expansion.
-    #[instrument(level = "debug", skip(self))]
-    pub fn fetch_events(
+/// A trait representing an identity manager that is able to submit user
+/// identities to a contract located on the blockchain.
+#[async_trait]
+pub trait IdentityManager {
+    /// Create and configure a new instance of the identity manager.
+    ///
+    /// # Arguments
+    ///
+    /// - `options`: The options used to configure the identity manager.
+    /// - `ethereum`: A connector for an ethereum-compatible blockchain.
+    async fn new(options: Options, ethereum: Ethereum) -> anyhow::Result<Self>
+    where
+        Self: Sized;
+
+    /// Returns the depth of the merkle tree managed by this `IdentityManager`.
+    fn tree_depth(&self) -> usize;
+
+    /// Returns the value used for a newly initialized merkle tree leaf.
+    fn initial_leaf_value(&self) -> Field;
+
+    /// Returns the group identifier associated with the identity manager.
+    fn group_id(&self) -> U256;
+
+    /// Returns the number of the latest block that is confirmed to have been
+    /// mined.
+    async fn confirmed_block_number(&self) -> Result<u64, EventError>;
+
+    /// Returns `true` if this `IdentityManager` acts via the manager address of
+    /// the on-chain contract it manages.
+    async fn is_owner(&self) -> anyhow::Result<bool>;
+
+    /// Registers the provided `identity_commitments` with the contract on
+    /// chain.
+    async fn register_identities(
+        &self,
+        identity_commitments: Vec<Field>,
+    ) -> Result<TransactionReceipt, TxError>;
+
+    /// Asserts that the provided `root` is the current root held by the
+    /// contract on the chain.
+    async fn assert_latest_root(&self, root: Field) -> anyhow::Result<()>;
+
+    /// Asserts that the provided `root` is a valid root.
+    ///
+    /// A valid root is one that has not expired based on the time since it was
+    /// inserted into the history of roots on chain.
+    async fn assert_valid_root(&self, root: Field) -> anyhow::Result<()>;
+
+    // TODO [Ara] Consider refactoring this to remove the stream when possible.
```

Comment:
No need for refactor. Once OZ Relay is merged https://github.com/worldcoin/signup-sequencer/pull/357, we can replace `fetch_events` with some other way of confirming transactions.